### PR TITLE
Refactor DecisionTreeRegressor to support XGBoost.

### PR DESCRIPTION
**Pull request by RishabhGarg108**
This PR is to refactor a few parts of the DecisionTreeRegressor to support the XGBoost algorithm. It should be merged only after #3011 is merged.

The following are the major refactorings to be done:-

- [x] Use instantiated `FitnessFunction` while training.
- [x] Move the `DecisionTreeRegressor::CalculatePrediction()` to `FitnessFunction::OutputLeafValue()`.

### CODE REVIEW

**Comment by RishabhGarg108** [comment]
@rcurtin @gmanlan Do take a look at these refactorings when you have time and let me know what you think about them. :)

**Review by rcurtin — COMMENTED** [review]
The general ideas here look good to me.  I have some comments about efficiency, but they should be easy to fix. :)  Do you think you can also add a line to the modified functions' documentation for each of the new parameters added?

It looks like one of the DecisionTreeRegressor tests failed:

```
/home/vsts/work/1/s/src/mlpack/tests/decision_tree_regressor_test.cpp:671
```

(That's from the 'Linux Plain' build.)  Maybe it is not very robust?  I'm not sure if the changes in this PR have affected that, but if you're willing to take a look into it and, e.g., try with many different random seeds, it would be good to be sure that the failure probability is low. :+1:

Nice work with this refactoring!  I think everything is basically ready to go here---my comments should hopefully be pretty quick to address.

File: src/mlpack/methods/decision_tree/mad_gain.hpp [hunk 1]
```diff
@@ -89,15 +89,37 @@ class MADGain
...
+          weightedSum);
+      return weightedSum / accWeights;
+    }
+    else
+    {
+      double sum;
+      Sum(responses, 0, responses.n_elem, sum);
+      return sum / responses.n_elem;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (RishabhGarg108) [reply src/mlpack/methods/decision_tree/mad_gain.hpp:122]: Yes, I agree on making the documentation more comprehensive. But, I am not sure if it will be worth it to add another template parameter for it. It is true that this is the same only for regression tree FitnessFunctions but it will be literally different for each and every loss function in gradient boosting trees and it will unnecessarily complicate the code there.

If we really want to reduce code duplication, then we can have a base `RegressionGain` class and implement it there. But, at this moment, I think that we don't need that too either. Maybe we can deal with it in future.

File: src/mlpack/tests/decision_tree_regressor_test.cpp [hunk 2]
```diff
@@ -203,12 +213,14 @@ TEST_CASE("AllCategoricalSplitSimpleSplitTest_", "[DecisionTreeRegressorTest]")
   AllCategoricalSplit<MSEGain>::AuxiliarySplitInfo aux;
 
   // Call the method to do the splitting.
-  const double bestGain = MSEGain::Evaluate<false>(responses, weights);
+  MSEGain Gain;
+  const double bestGain = Gain.Evaluate<false>(responses, weights);
   const double gain = AllCategoricalSplit<MSEGain>::SplitIfBetter<false>(
-      bestGain, predictor, 2, responses, weights, 3, 1e-7, splitInfo, aux);
+      bestGain, predictor, 2, responses, weights, 3, 1e-7, splitInfo, aux,
+      Gain);
```
[the remark this answers is not in the document]
Comment (RishabhGarg108) [reply src/mlpack/tests/decision_tree_regressor_test.cpp:220]: Sure. I will fix it to adhere to the style guide 👍

File: src/mlpack/methods/decision_tree/all_categorical_split.hpp [hunk 3]
```diff
@@ -103,7 +103,8 @@ class AllCategoricalSplit
       const size_t minimumLeafSize,
       const double minimumGainSplit,
       double& splitInfo,
-      AuxiliarySplitInfo& aux);
+      AuxiliarySplitInfo& aux,
+      FitnessFunction& fitnessFunction);
```
Comment (rcurtin) [inline src/mlpack/methods/decision_tree/all_categorical_split.hpp:107]: Can you add documentation for the `fitnessFunction` parameter to the function documentation?  (This applies for the other functions where you added it too. :+1:)

File: src/mlpack/methods/decision_tree/mad_gain.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (rcurtin) [reply src/mlpack/methods/decision_tree/mad_gain.hpp:124]: Agreed---let's avoid the extra template parameter for now then. :+1:

File: src/mlpack/methods/decision_tree/mse_gain.hpp [hunk 4]
```diff
@@ -41,10 +41,10 @@ class MSEGain
    * @param end End index.
    */
   template<bool UseWeights, typename VecType, typename WeightVecType>
-  static double Evaluate(const VecType& values,
```
Comment (rcurtin) [inline src/mlpack/methods/decision_tree/mse_gain.hpp:44]: Was there a reason to change these from `static`?  I think they should work just fine as they were---let me know if I overlooked something.

**Review by rcurtin — COMMENTED** [review]
Looking good---just a couple comments to address and I think this is ready for merge. :+1:

File: src/mlpack/methods/decision_tree/mse_gain.hpp (hunk 4 above)
Comment (RishabhGarg108) [reply src/mlpack/methods/decision_tree/mse_gain.hpp:44]: I think we can remove this function and move its content to the overload just below it. Because we are using this function only in line https://github.com/RishabhGarg108/mlpack-1/blob/a4081bf0c8029f70b1a9d0a36851d3285fccd43d/src/mlpack/methods/decision_tree/mse_gain.hpp#L95.

Regarding it being non-static, I kept it this way because the place at which this function is called ([here](https://github.com/RishabhGarg108/mlpack-1/blob/a4081bf0c8029f70b1a9d0a36851d3285fccd43d/src/mlpack/methods/decision_tree/decision_tree_regressor_impl.hpp#L625)), we are already using an instantiated `FitnessFunction`. So it won't matter if it is static or not.

If you like then I can make it static back again. No harm in that :) Let me know what do you think 👍

Comment (rcurtin) [reply src/mlpack/methods/decision_tree/mse_gain.hpp:44]: Yeah, you are right, it doesn't matter, but I do think it keeps the tests a bit simpler.  So I would suggest changing it, but, in the end, up to you. :+1:

**Review by rcurtin — APPROVED** [review]
Looks good to me!  Thanks for making the changes to add documentation. :+1:
